### PR TITLE
Added a generic qr code reader

### PR DESCRIPTION
**Pull request by TomAFrench**
## Issue number

#238
## Description

I've replaced the current qr reader provided by web3 with [react-qr-reader](https://www.npmjs.com/package/react-qr-reader)

I need some help from the Kickback team before I can do further testing (details below)

## List of features added/changed

- stripped out old references to the qrcoderesolver
- Added in react-qr-reader

## How Has This Been Tested?

It's a bit of a pain testing this as the page it's on is hidden behind an admin permissions test so I don't think I can really see what the current behaviour is except what I can see from the code.

I've pulled `EventFilters` onto another page which allowed me to make sure that addresses are properly scanned and it fails gracefully without a camera but I can't go much more beyond that atm.

Would it be possible for me to be added as an admin on an event on the dev version of the site? 

- [x] Can scan addresses
- [x] Handles malformed addresses
- [x] Scan QR button is greyed out if camera permissions are denied
- [ ] Mobile testing
## Screenshots (if appropriate):

This shows the qr scanner which has been placed on the `/create` page

alt

Should camera permissions be denied or a fatal error occur (such as not having a camera at all) then the scan qr button is greyed out and disabled.

alt

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My code implements all the required features.

### CODE REVIEW

**Comment by makoto** [comment]
Hey @TomAFrench if you point your frontend to kovan, you can create an event on your own from /create (and password is blank). If for some reason you are having problem, then send me your ETH address soo that I can add you to one of existing events

**Comment by TomAFrench** [comment]
Hey @makoto, I'm afraid I don't have kovan ETH (the faucet alternates between not having enough funds to send any and telling me to wait 24hrs) so I can't deploy my own event contact.

My address is tomfrench.eth. If you have some spare, could you send me enough kovan ETH for testing this please?

**Comment by TomAFrench** [comment]
My god, I must have been pretty tired when I wrote this PR because needing admin access to get around the HTTPS issue makes no sense. I'll still need to set up a server with a valid certificate myself.

**Comment by makoto** [comment]
Hi, I sent you 3 Kovan ETH and 10 test DAI.

**Comment by TomAFrench** [comment]
Cool, I've just tested it and things are hooked up correctly in terms of searching for attendees.

All that remains is I need to host the somewhere public for testing for mobile as apps won't use my CA and then testing the address formats for popular wallets (for example coinbase wallet shows the address prefixed by "ethereum:")

**Comment by makoto** [comment]
You can use https://pr245-app-wearekickback.surge.sh/

**Comment by makoto** [comment]
QR code from desktop seems working fine but I got error when reading from mobile (tested with metamask mobile).

Screenshot_20200110-222529

I do not think it is due to the camera quality of my phone nor metamask webview because when I used the QR code reader of https://xdai.io/ it scanned okay.

You may want to check the source of burner wallet as it looks like it is using two different qr reader libraries so they may be doing some extra ways to ensure that the QR code works.

https://github.com/austintgriffith/burner-wallet/blob/bc4a00c79bba95342be0ee0a944b599b9b616f2d/src/components/SendByScan.js#L152

**Comment by TomAFrench** [comment]
> QR code from desktop seems working fine but I got error when reading from mobile (tested with metamask mobile).
> 
> I do not think it is due to the camera quality of my phone nor metamask webview because when I used the QR code reader of https://xdai.io/ it scanned okay. 

Hey, It's strange you're having issues with scanning using metamask as it seems to work fine for me. Based on the error message you got it seems like it could read the code but it just didn't conform to the address format it expects. At the minute it expects just a simple "0x..." but I know at least one wallet which deviates from this so I'll be adding a little more logic to handle these. Do you mind sharing the QR code you're using for testing?

> You may want to check the source of burner wallet as it looks like it is using two different qr reader libraries so they may be doing some extra ways to ensure that the QR code works.

Yep, I've done some work on the burner wallet in the past. The second qr reader is for the the situation where the browser can't directly use the camera, instead the event operator would have to manually take a picture of the qr code and then open that file in the browser. It's not a great UX experience.

**Comment by TomAFrench** [comment]
Successful tests checklist

### Desktop
- [ ] Chrome
- [x] Brave
- [ ] Opera
### Mobile
- [ ] Status.im
- [ ] Trust Wallet
- [ ] Metamask Mobile
- [x] Coinbase wallet
- [ ] Opera

**Comment by TomAFrench** [comment]
@makoto Can you push this

**Comment by TomAFrench** [comment]
Is there a list of wallets which the previous qr reader supported?

**Comment by makoto** [comment]
@TomAFrench How difficult to add a feature to show warning message if certain browsers do not allow camera access?

**Comment by makoto** [comment]
@TomAFrench How difficult to add a feature to show warning message if certain browsers do not allow camera access?

File: src/components/SingleEvent/EventFilters.js [hunk 1]
```diff
@@ -90,22 +113,34 @@ class EventFilters extends Component {
...
+                <CenteredQrReader
+                  delay={400} // delay = false stops scanning
+                  onError={err => {
+                    console.log(err)
+                    this.setState({
+                      scannerOn: false,
+                      scanError: 'FATAL ERROR'
+                    })
+                    this._disableCamera()
+                    setTimeout(() => this.setState({ scanError: false }), 3000)
```
Comment (jefflau) [inline src/components/SingleEvent/EventFilters.js]: Can we add a comment to explain the logic behind this timeout? Seems like the error just pops up 3 seconds and then disappears? If so we can also add a variable to explain 3000, such as `const TIME_ERROR_APPEARS` or something similar

File: src/components/SingleEvent/EventFilters.js [hunk 2]
```diff
@@ -22,28 +21,52 @@ const Filter = styled('div')`
...
+  }
+
+  _disableCamera = () => {
+    this.setState({ cameraUnavailable: true })
+  }
+
+  _scan = data => {
+    console.log(data)
+    if (!data) {
+      console.log('No data')
```
Comment (jefflau) [inline src/components/SingleEvent/EventFilters.js]: Replace logs with a message to return to the user.

Remove the other logs unless you think they are useful for something other than debugging your own code

File: src/components/SingleEvent/EventFilters.js (hunk 1 above)
Comment (TomAFrench) [reply src/components/SingleEvent/EventFilters.js]: Sure thing

> @TomAFrench How difficult to add a feature to show warning message if certain browsers do not allow camera access?

Not hard at all. I'll add it

**Comment by TomAFrench** [comment]
> @TomAFrench How difficult to add a feature to show warning message if certain browsers do not allow camera access?

Not hard at all. I'll add it

**Comment by TomAFrench** [comment]
> If you are feeling up to it for bonus points, you could refactor EventFilters component into using hooks.

I've done this refactoring. It's much cleaner now.

File: src/components/SingleEvent/EventFilters.js [hunk 3]
```diff
@@ -33,118 +33,114 @@ const CenteredQrReader = styled(QrReader)`
   width: 75%;
 `
 
-class EventFilters extends Component {
-  constructor(props) {
-    super(props)
+const decodeQR = data => {
+  console.log(data)
```
Comment (jefflau) [inline src/components/SingleEvent/EventFilters.js]: remove log

Comment (TomAFrench) [reply src/components/SingleEvent/EventFilters.js]: Ah whoopsie. Sorry that slipped through

**Pull request merged**